### PR TITLE
Training/2021.10/070 subprograms

### PR DESCRIPTION
**Pull request by leogermond**

### CODE REVIEW

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 1]
```diff
@@ -77,136 +76,118 @@ A Little "Preaching" About Names
...
+   - **Declaration** and **specification** are used synonymously
 
-   - Provide the subprogram specification
-   - *Declaration* and *specification* are often used synonymously
+* Specification may be required
 
-* Required in some circumstances
+   - Package specification
+   - Recursion
+   - Optional otherwise
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:95]: This last bullet isn't grammatically the same as the other two - maybe put parentheses around it?

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 2]
```diff
@@ -532,68 +440,47 @@ Why Read Mode `out` Parameters?
...
-
 ---------------------------------
 Modes' Requirements for Actuals
 ---------------------------------
 
-* Use of variables versus expressions for actuals
 * Modes :ada:`in out` and :ada:`out`
 
-   - Variables must be used since actual may/will be altered
+   - Must **not** use expressions
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:449]: Should be more like "Must use writable objects (no expressions)"

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 3]
```diff
@@ -701,95 +587,76 @@ Implementation-Dependent Types
 Unconstrained Formal Parameters
 ---------------------------------
 
-* Take bounds from actual parameters
+* Take bounds from **actual** parameters
+* Also applies to :ada:`return` values
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:591]: No - return object is bounded by the function (caller typically uses the function to initialize an unconstrained object)

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 4]
```diff
@@ -1669,9 +1355,9 @@ Summary
 Summary
 ---------
 
-* Procedures are abstractions for actions
-* Functions are abstractions for value computations
-* Functions may return composite values
+* :ada:`procedure` are abstractions for actions
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:1358]: if you're going to use singular, then "is an abstraction ..."

**Review by frank-at-adacore — CHANGES_REQUESTED** [review]

File: courses/fundamentals_of_ada/070_subprograms.rst (hunk 3 above)
Comment (leogermond) [reply courses/fundamentals_of_ada/070_subprograms.rst:591]: Return as in formal return, which can be declared unconstrained, rewriting that

File: courses/fundamentals_of_ada/070_subprograms.rst (hunk 4 above)
Comment (leogermond) [reply courses/fundamentals_of_ada/070_subprograms.rst:1358]: Done

**Review by frank-at-adacore — DISMISSED** [review]

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 5]
```diff
@@ -1151,90 +995,49 @@ Expression Functions
 
    Ada 2012
 
-* Shorthand for declaring functions whose implementations are only "expressions"
+* Functions whose implementations are pure expressions
 
-   - Only the returned value appears
+   - No body is allowed
```
[the remark this answers is not in the document]
Comment (leogermond) [reply courses/fundamentals_of_ada/070_subprograms.rst:1000]: Ok I see your point, expression-functions provide a shorthand to a body declaration, let's clarify that no other completion is allowed indeed.

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 6]
```diff
@@ -77,68 +76,58 @@ A Little "Preaching" About Names
...
+   - **Declaration** and **specification** are used synonymously
 
-   - Provide the subprogram specification
-   - *Declaration* and *specification* are often used synonymously
+* Specification may be required in
 
-* Required in some circumstances
+   - Package specification
+   - Recursion
+   - (Optional otherwise)
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:95]: There is at least two other non-optional (private child subprograms, generic subprograms). I would either complete the list (and I'm not sure I know what the full list would be) or just leave the original (perhaps removing "You'll see them")

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 7]
```diff
@@ -77,68 +76,58 @@ A Little "Preaching" About Names
...
+.. code:: Ada
 
-      .. code:: Ada
+   function F (X : Real) return Real;
 
-         function F (X : Real) return Real is
-         begin
-           return X + 3.0 * X;
-         end F;
+* Same as :ada:`procedure`
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:127]: You say "same as" and then you switch names around.
e.g procedure name is a "defining_name" but function name can be "operator" or "identifier"
same with "parameter_profile" and "formal_part"

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 8]
```diff
@@ -148,74 +137,63 @@ Procedure Declaration Syntax (Simplified)
...
-   - Separate declarations not required to define subprograms
+   - Mostly **semantic** check
+   - But parameters **must** have same name
+
+.. code:: Ada
+
+   procedure P (J, K : Integer)
+   procedure P (J : Integer; K : Integer)
+   procedure P (J, K : in Integer)
+   -- Invalid
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:195]: This is not invalid - it's just an (overloaded) procedure

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 9]
```diff
@@ -373,48 +299,33 @@ Examples
 Subprogram Parameter Terminology
 ----------------------------------
 
-* *Actual* parameters are values passed to a call
+* **Actual** parameters are values passed to a call
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:302]: Not sure about this. I typically assume bold to be emphasis, and italics to be definition/explanation. We're not emphasizing here, we're just describing what Actual and Formal mean

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 10]
```diff
@@ -713,95 +602,78 @@ Implementation-Dependent Types
...
-* Take bounds from actual parameters
+* Take bounds from **actual** parameters
+* Also applies to formal :ada:`return` specification
+
+    + Actual :ada:`return` is constrained in the body
 
 .. code:: Ada
 
    type Vector is array (Positive range <>) of Real;
+   procedure Print (V : Vector);
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:613]: Maybe switch "V" to "Formal" so the comments later are easier to understand?

File: courses/fundamentals_of_ada/070_subprograms.rst [hunk 11]
```diff
@@ -1318,56 +1114,21 @@ Mode `out` Risk for Scalars
 * Any effect upon external objects or external environment
 
    - Typically alteration of non-local variables or states
+   - Can cause hard-to-debug errors
+   - Not legal in SPARK
```
Comment (frank-at-adacore) [inline courses/fundamentals_of_ada/070_subprograms.rst:1118]: This is only true for functions - your example shows a procedure

**Review by frank-at-adacore — CHANGES_REQUESTED** [review]

File: courses/fundamentals_of_ada/070_subprograms.rst (hunk 6 above)
Comment (leogermond) [reply courses/fundamentals_of_ada/070_subprograms.rst:95]: We provide example later in the slide for those two. So I'm adding back the original, but on top of them

File: courses/fundamentals_of_ada/070_subprograms.rst (hunk 11 above)
Comment (leogermond) [reply courses/fundamentals_of_ada/070_subprograms.rst]: Good catch, it's been in the material for a while.

**Review by leogermond — COMMENTED** [review]
Ok for most, I disagree on the overload vs redefinition remark, and GNAT seems to agree with me.

Also how about using the :dfn: role (https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-dfn) for defintiions? This would have the added benefit of permitting some tracking of where exactly the concepts are introduced

**Comment by frank-at-adacore** [comment]
I guess the overload thing makes sense - my bad!
If :dfn: works with Pandoc, then I'm good with that

**Review by frank-at-adacore — APPROVED** [review]

**Pull request merged**